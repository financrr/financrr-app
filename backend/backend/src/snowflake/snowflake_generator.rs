use crate::snowflake::snowflake_error::SnowflakeGeneratorError;
use crate::time::timestamp::get_epoch_millis;
use std::sync::atomic::{AtomicU64, Ordering};

pub(crate) const SNOWFLAKE_EPOCH: i64 = 1_705_247_483_000;

pub(crate) const NODE_ID_BITS: u8 = 10;
pub(crate) const SEQUENCE_BITS: u8 = 12;

pub(crate) const TIMESTAMP_SHIFT: u8 = NODE_ID_BITS + SEQUENCE_BITS;

pub(crate) const MAX_NODE_ID: u64 = (1 << NODE_ID_BITS) - 1;
pub(crate) const MAX_SEQUENCE: u64 = (1 << SEQUENCE_BITS) - 1;

#[derive(Debug)]
pub(crate) struct SnowflakeGenerator {
    node_id: u64,
    last_timestamp: AtomicU64,
    sequence: AtomicU64,
}

impl SnowflakeGenerator {
    pub(crate) fn new(node_id: u64) -> Self {
        Self {
            node_id,
            last_timestamp: AtomicU64::new(0),
            sequence: AtomicU64::new(0),
        }
    }

    pub(crate) fn next_id(&self) -> Result<i64, SnowflakeGeneratorError> {
        let mut current_timestamp = self.timestamp();
        let last_timestamp = self.last_timestamp.load(Ordering::SeqCst);

        if current_timestamp < last_timestamp {
            return Err(SnowflakeGeneratorError::InvalidSystemClock);
        }

        let mut sequence = self.sequence.load(Ordering::SeqCst);

        if current_timestamp == last_timestamp {
            sequence = (sequence + 1) & MAX_SEQUENCE;
            if sequence == 0 {
                current_timestamp = self.wait_for_next_millis(current_timestamp, last_timestamp)?;
            }
        } else {
            sequence = 0;
        }

        self.last_timestamp.store(current_timestamp, Ordering::SeqCst);
        self.sequence.store(sequence, Ordering::SeqCst);

        Ok(((current_timestamp << (NODE_ID_BITS + SEQUENCE_BITS)) | (self.node_id << SEQUENCE_BITS) | sequence) as i64)
    }

    fn timestamp(&self) -> u64 {
        (get_epoch_millis() - (SNOWFLAKE_EPOCH as u128)) as u64
    }

    fn wait_for_next_millis(
        &self,
        mut current_timestamp: u64,
        last_timestamp: u64,
    ) -> Result<u64, SnowflakeGeneratorError> {
        while current_timestamp == last_timestamp {
            current_timestamp = self.timestamp();
        }

        Ok(current_timestamp)
    }
}
