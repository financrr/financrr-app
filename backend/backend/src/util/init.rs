use std::fmt::Debug;
use std::process;

use tracing::error;

pub(crate) fn expect_or_exit<T, E: Debug>(result: Result<T, E>, message: &str) -> T {
    match result {
        Ok(value) => value,
        Err(err) => {
            error!("{}: {:?}", message, err);
            process::exit(1);
        }
    }
}
