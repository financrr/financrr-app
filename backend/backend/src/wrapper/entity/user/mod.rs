use actix_web::dev::Payload;
use actix_web::{FromRequest, HttpRequest};
use futures_util::future::LocalBoxFuture;
use sea_orm::{EntityName, EntityTrait};
use serde::{Deserialize, Serialize};
use time::OffsetDateTime;
use utoipa::ToSchema;

use dto::Credentials;
use entity::prelude::User as DbUser;
use entity::user;
use entity::user::Model;

use crate::api::error::api::ApiError;
use crate::database::entity::{count, find_one, find_one_or_error, insert};
use crate::permission_impl;
use crate::util::auth::extract_bearer_token;
use crate::wrapper::entity::session::Session;
use crate::wrapper::entity::user::dto::UserRegistration;
use crate::wrapper::entity::{TableName, WrapperEntity};
use crate::wrapper::permission::{Permission, Permissions};
use crate::wrapper::types::phantom::{Identifiable, Phantom};

pub(crate) mod dto;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
pub(crate) struct User {
    pub(crate) id: i32,
    pub(crate) username: String,
    pub(crate) email: Option<String>,
    pub(crate) display_name: Option<String>,
    #[serde(with = "time::serde::rfc3339")]
    pub(crate) created_at: OffsetDateTime,
    pub(crate) is_admin: bool,
}

impl User {
    pub(crate) async fn exists(id: i32) -> Result<bool, ApiError> {
        Ok(count(user::Entity::find_by_id(id)).await? > 0)
    }

    pub(crate) async fn authenticate(credentials: &Credentials) -> Result<Self, ApiError> {
        let user = find_one(DbUser::find_by_username(credentials.username.as_str())).await;
        match user {
            Ok(Some(user)) => {
                if user.verify_password(credentials.password.as_bytes()).unwrap_or(false) {
                    Ok(Self::from(user))
                } else {
                    Err(ApiError::InvalidCredentials())
                }
            }
            _ => Err(ApiError::InvalidCredentials()),
        }
    }

    pub(crate) async fn register(registration: UserRegistration) -> Result<Self, ApiError> {
        match user::ActiveModel::register(
            registration.username,
            registration.email,
            registration.display_name,
            registration.password,
        ) {
            Ok(user) => {
                let model = insert(user).await?;
                let user = Self::from(model);
                user.add_permission(user.id, Permissions::all()).await?;

                Ok(user)
            }
            Err(e) => Err(ApiError::from(e)),
        }
    }
}

permission_impl!(User);

impl Identifiable for User {
    async fn find_by_id(id: i32) -> Result<Self, ApiError>
    where
        Self: Sized,
    {
        find_one_or_error(user::Entity::find_by_id(id), "User").await.map(Self::from)
    }
}

impl TableName for User {
    fn table_name() -> &'static str {
        user::Entity.table_name()
    }
}

impl WrapperEntity for User {
    fn get_id(&self) -> i32 {
        self.id
    }
}

impl FromRequest for User {
    type Error = ApiError;
    type Future = LocalBoxFuture<'static, Result<Self, Self::Error>>;

    fn from_request(req: &HttpRequest, _payload: &mut Payload) -> Self::Future {
        let req = req.clone();
        Box::pin(async move {
            // get bearer token from request
            let token = extract_bearer_token(&req)?;
            let user_id = Session::find_user_id(token).await?;

            Self::find_by_id(user_id).await
        })
    }
}

impl FromRequest for Phantom<User> {
    type Error = ApiError;
    type Future = LocalBoxFuture<'static, Result<Self, Self::Error>>;

    fn from_request(req: &HttpRequest, _payload: &mut Payload) -> Self::Future {
        let req = req.clone();
        Box::pin(async move {
            let token = extract_bearer_token(&req)?;
            let user_id = Session::find_user_id(token).await?;
            User::exists(user_id).await?;

            Ok(Self::new(user_id))
        })
    }
}

impl From<user::Model> for User {
    fn from(value: Model) -> Self {
        Self {
            id: value.id,
            username: value.username,
            email: value.email,
            display_name: value.display_name,
            created_at: value.created_at,
            is_admin: value.is_admin,
        }
    }
}
