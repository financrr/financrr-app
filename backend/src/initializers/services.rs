use crate::services::configure_services;
use async_trait::async_trait;
use axum::Router as AxumRouter;
use loco_rs::app::AppContext;
use loco_rs::prelude::Initializer;

pub struct ServicesInitializer;

#[async_trait]
impl Initializer for ServicesInitializer {
    fn name(&self) -> String {
        "Services".to_string()
    }

    async fn after_routes(&self, router: AxumRouter, ctx: &AppContext) -> loco_rs::Result<AxumRouter> {
        Ok(configure_services(router, ctx).await?)
    }
}
