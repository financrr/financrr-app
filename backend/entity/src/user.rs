//! `SeaORM` Entity, @generated by sea-orm-codegen 1.0.0-rc.4

use sea_orm::entity::prelude::*;
use sea_orm::ActiveValue::Set;
use serde::{Deserialize, Serialize};

use crate::error::EntityError;
use crate::utility::hashing::hash_string;
use crate::utility::time::get_now;

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "user")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: i32,
    #[sea_orm(column_type = "Text", unique)]
    pub username: String,
    #[sea_orm(column_type = "Text", nullable, unique)]
    pub email: Option<String>,
    #[sea_orm(column_type = "Text", nullable)]
    pub display_name: Option<String>,
    #[sea_orm(column_type = "Text")]
    pub password: String,
    pub created_at: TimeDateTimeWithTimeZone,
    pub is_admin: bool,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(has_many = "super::budget::Entity")]
    Budget,
    #[sea_orm(has_many = "super::currency::Entity")]
    Currency,
    #[sea_orm(has_many = "super::permissions::Entity")]
    Permissions,
    #[sea_orm(has_many = "super::session::Entity")]
    Session,
}

impl Related<super::budget::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Budget.def()
    }
}

impl Related<super::currency::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Currency.def()
    }
}

impl Related<super::permissions::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Permissions.def()
    }
}

impl Related<super::session::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Session.def()
    }
}

impl ActiveModelBehavior for ActiveModel {}

impl Model {
    pub fn verify_password(&self, password: &[u8]) -> Result<bool, EntityError> {
        argon2::verify_encoded(&self.password, password).map_err(EntityError::HashingFailed)
    }
}

impl ActiveModel {
    pub fn register(
        username: String,
        email: Option<String>,
        display_name: Option<String>,
        password: String,
    ) -> Result<Self, EntityError> {
        let hashed_password = hash_string(&password)?;

        Ok(Self {
            id: Default::default(),
            username: Set(username),
            email: Set(email),
            display_name: Set(display_name),
            password: Set(hashed_password),
            created_at: Set(get_now()),
            is_admin: Set(false),
        })
    }
}

impl Entity {
    pub fn find_by_username(username: &str) -> Select<Self> {
        Self::find().filter(Column::Username.eq(username.to_string()))
    }
}
