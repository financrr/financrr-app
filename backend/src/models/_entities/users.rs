//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.3

use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "users")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub id: i64,
    #[sea_orm(column_type = "Text", unique)]
    pub email: String,
    #[sea_orm(column_type = "Text")]
    pub name: String,
    pub flags: i32,
    #[sea_orm(column_type = "Text")]
    pub password: String,
    #[sea_orm(column_type = "Text", nullable)]
    pub reset_token: Option<String>,
    pub reset_sent_at: Option<DateTimeWithTimeZone>,
    #[sea_orm(column_type = "Text", nullable)]
    pub email_verification_token: Option<String>,
    pub email_verification_sent_at: Option<DateTimeWithTimeZone>,
    pub email_verified_at: Option<DateTimeWithTimeZone>,
    pub created_at: DateTimeWithTimeZone,
    pub updated_at: DateTimeWithTimeZone,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(has_many = "super::categories::Entity")]
    Categories,
    #[sea_orm(has_many = "super::currencies::Entity")]
    Currencies,
    #[sea_orm(has_many = "super::sessions::Entity")]
    Sessions,
    #[sea_orm(has_many = "super::tags::Entity")]
    Tags,
    #[sea_orm(has_many = "super::user_permissions::Entity")]
    UserPermissions,
}

impl Related<super::categories::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Categories.def()
    }
}

impl Related<super::currencies::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Currencies.def()
    }
}

impl Related<super::sessions::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Sessions.def()
    }
}

impl Related<super::tags::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Tags.def()
    }
}

impl Related<super::user_permissions::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::UserPermissions.def()
    }
}
