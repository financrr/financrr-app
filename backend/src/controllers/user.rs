use crate::error::app_error::AppError;
use crate::models::users::{Model, RegisterParams};
use crate::services::snowflake_generator::SnowflakeGenerator;
use crate::services::user_verification::UserVerificationService;
use crate::views::user::UserResponse;
use axum::extract::State;
use axum::http::StatusCode;
use axum::routing::post;
use axum::{debug_handler, Extension, Json};
use loco_rs::app::AppContext;
use loco_rs::prelude::Routes;
use sea_orm::IntoActiveModel;
use serde::Deserialize;
use utoipa::ToSchema;
use validator::{Validate, ValidationError, ValidationErrors};

#[derive(Debug, Deserialize, ToSchema)]
pub struct VerifyParams {
    pub token: String,
}

/// Registers a new User.
#[utoipa::path(post,
    path = "/api/v1/users/register",
    tag = "User",
    responses(
        (status = StatusCode::CREATED, description = "Successfully registered a new User.", content_type="application/json", body = UserResponse),
        (status = StatusCode::BAD_REQUEST, description = "Validation error.", content_type="application/json", body = AppError),
    )
)]
#[debug_handler]
async fn register(
    State(ctx): State<AppContext>,
    Extension(user_verification_service): Extension<UserVerificationService>,
    Extension(snowflake_generator): Extension<SnowflakeGenerator>,
    Json(params): Json<RegisterParams>,
) -> Result<(StatusCode, Json<UserResponse>), AppError> {
    params.validate()?;
    if !Model::is_email_unique(&ctx.db, &params.email).await? {
        let mut errors = ValidationErrors::new();
        errors.add("email", ValidationError::new("Email already exists."));

        return Err(errors.into());
    }

    let res = Model::create_with_password(&ctx.db, &snowflake_generator, &params).await?;
    let active_model = res.into_active_model();

    let model = user_verification_service
        .send_verification_email_or_verify_user(active_model)
        .await?;

    Ok((StatusCode::CREATED, Json(UserResponse::from(model))))
}

/// Verifies a User with the given token.
#[utoipa::path(post,
    path = "/api/v1/users/verify",
    tag = "User",
    responses(
        (status = StatusCode::OK, description = "Successfully verified a User.", content_type="application/json", body = UserResponse),
        (status = StatusCode::BAD_REQUEST, description = "Invalid verification token.", content_type="application/json", body = AppError),
    )
)]
#[debug_handler]
async fn verify(
    State(ctx): State<AppContext>,
    Json(params): Json<VerifyParams>,
) -> Result<(StatusCode, Json<UserResponse>), AppError> {
    let user = Model::find_by_verification_token(&ctx.db, &params.token).await?;
    match user {
        None => Err(AppError::InvalidVerificationToken()),
        Some(user) => {
            let active_model = user.into_active_model();
            let user = active_model.verified(&ctx.db).await?;

            Ok((StatusCode::OK, Json(UserResponse::from(user))))
        }
    }
}

pub fn routes() -> Routes {
    Routes::new()
        .prefix("/users")
        .add("/register", post(register))
        .add("/verify", post(verify))
}
