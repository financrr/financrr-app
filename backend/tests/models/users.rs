use crate::helpers::init::init_test;
use crate::helpers::users::{clean_up_user_model, create_user_with_email, generate_test_user};
use financrr::controllers::user::RegisterParams;
use financrr::error::app_error::AppResult;
use financrr::services::snowflake_generator::SnowflakeGeneratorInner;
use financrr::services::Service;
use financrr::{app::App, models::users::Model};
use insta::assert_debug_snapshot;
use loco_rs::testing;
use sea_orm::IntoActiveModel;
use serial_test::serial;

macro_rules! configure_insta {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        settings.set_snapshot_suffix("users");
        let _guard = settings.bind_to_scope();
    };
}

#[tokio::test]
#[serial]
async fn can_create_with_password() {
    init_test!();

    let boot = testing::boot_test::<App>().await.unwrap();

    let params = RegisterParams {
        email: "test@financrr.dev".to_string(),
        password: "Password1234".to_string(),
        name: "Test Account".to_string(),
    };
    let snowflake_generator = SnowflakeGeneratorInner::new_arc(&boot.app_context).await.unwrap();

    let res = Model::create_with_password(&boot.app_context.db, &snowflake_generator, &params).await;

    insta::with_settings!({
        filters => clean_up_user_model()
    }, {
        assert_debug_snapshot!(res);
    });
}

#[tokio::test]
#[serial]
async fn handle_create_with_password_with_duplicate() {
    init_test!();

    let boot = testing::boot_test::<App>().await.unwrap();

    let snowflake_generator = SnowflakeGeneratorInner::get_arc(&boot.app_context).await.unwrap();
    let _ = create_user_with_email(&boot.app_context, "duplicate_user@financrr.test").await;

    let new_user: AppResult<Model> = Model::create_with_password(
        &boot.app_context.db,
        &snowflake_generator,
        &RegisterParams {
            email: "duplicate_user@financrr.test".to_string(),
            password: "Password1234".to_string(),
            name: "Test Account".to_string(),
        },
    )
    .await;

    insta::with_settings!({
        filters => clean_up_user_model()
    }, {
        assert_debug_snapshot!(new_user);
    });
}

#[tokio::test]
#[serial]
async fn can_find_by_email() {
    init_test!();

    let boot = testing::boot_test::<App>().await.unwrap();

    let user = create_user_with_email(&boot.app_context, "user1@financrr.test").await;

    let existing_user = Model::find_by_email(&boot.app_context.db, user.email.as_str()).await;
    let non_existing_user_results = Model::find_by_email(&boot.app_context.db, "non-existing@financrr.test").await;

    assert_debug_snapshot!(existing_user);
    assert_debug_snapshot!(non_existing_user_results);
}

#[tokio::test]
#[serial]
async fn can_verification_token() {
    init_test!();

    let boot = testing::boot_test::<App>().await.unwrap();

    let user = Model::find_by_id(&boot.app_context.db, 1).await.unwrap();

    assert!(user.email_verification_sent_at.is_none());
    assert!(user.email_verification_token.is_none());

    assert!(user
        .into_active_model()
        .set_email_verification_sent(&boot.app_context.db)
        .await
        .is_ok());

    let user = Model::find_by_id(&boot.app_context.db, 1).await.unwrap();

    assert!(user.email_verification_sent_at.is_some());
    assert!(user.email_verification_token.is_some());
}

#[tokio::test]
#[serial]
async fn can_set_forgot_password_sent() {
    init_test!();

    let boot = testing::boot_test::<App>().await.unwrap();
    let user = generate_test_user(&boot.app_context).await;
    let user_id = user.id;

    assert!(user.reset_sent_at.is_none());
    assert!(user.reset_token.is_none());

    assert!(user
        .into_active_model()
        .set_forgot_password_sent(&boot.app_context.db)
        .await
        .is_ok());

    let user = Model::find_by_id(&boot.app_context.db, user_id).await.unwrap();

    assert!(user.reset_sent_at.is_some());
    assert!(user.reset_token.is_some());
}

#[tokio::test]
#[serial]
async fn can_verified() {
    init_test!();

    let boot = testing::boot_test::<App>().await.unwrap();

    let user = Model::find_by_id(&boot.app_context.db, 1).await.unwrap();

    assert!(user.email_verified_at.is_none());

    assert!(user.into_active_model().verified(&boot.app_context.db).await.is_ok());

    let user = Model::find_by_id(&boot.app_context.db, 1).await.unwrap();

    assert!(user.email_verified_at.is_some());
}

#[tokio::test]
#[serial]
async fn can_reset_password() {
    init_test!();

    let boot = testing::boot_test::<App>().await.unwrap();

    let user = Model::find_by_id(&boot.app_context.db, 1).await.unwrap();

    assert!(user.verify_password("12341234"));

    assert!(user
        .clone()
        .into_active_model()
        .reset_password(&boot.app_context.db, "new-password")
        .await
        .is_ok());

    assert!(Model::find_by_id(&boot.app_context.db, 1)
        .await
        .unwrap()
        .verify_password("new-password"));
}
