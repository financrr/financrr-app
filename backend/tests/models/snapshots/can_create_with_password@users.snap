---
source: tests/models/users.rs
assertion_line: 44
expression: res
snapshot_kind: text
---
Ok(
    Model {
        id: ID
        email: "test@financrr.dev",
        name: "Test Account",
        flags: 2,
        password: "PASSWORD",
        reset_token: None,
        reset_sent_at: None,
        email_verification_token: None,
        email_verification_sent_at: None,
        email_verified_at: None,
        created_at: DATE,
        updated_at: DATE,
    },
)
